Cython/Compiler/ExprNodes.py

```diff
@@ -302,6 +302,80 @@ def translate_double_cpp_exception(code, pos, lhs_type, lhs_code, rhs_code, lhs_
     code.putln('}')
 
 
+def analyse_thread_safety(node, env):
+    """
+    Either leaves "node.needs_threadsafe_access" or sets it
+    to the scope that it should be locked with.
+    """
+    if env.nogil:
+        return  # we only do thread safety with the GIL
+    if env.is_c_class_scope:
+        # evaluation of C class scope is assumed not to
+        # need guarding
+        return
+    directive = env.directives['threadsafe_variable_access']
+    if directive == 'off':
+        return  # the user has told us it doesn't need thread safety
+    if directive == 'refcounted' and not node.type.needs_refcounting:
+        return
+    entry = node.entry
+    if entry is None:
+        return
+    if entry.type.is_cfunction:
+        # C functions are immutable so can be special-cased out
+        return
+    if entry.scope.scope_needs_threadsafe_lookup:
+        node.needs_threadsafe_access = True
+        node.threadsafety_scope = entry.scope
+        return
+    if entry.scope.is_module_scope and entry.is_cglobal:
+        node.needs_threadsafe_access = True
+        node.threadsafety_scope = entry.scope
+        return
+    if entry.from_closure and entry.outer_entry.scope.scope_needs_threadsafe_lookup:
+        node.needs_threadsafe_access = True
+        node.threadsafety_scope = entry.outer_entry.scope
+        return
+    if not entry.scope.is_local_scope:
+        # Local scope has complicated rules due to parallel blocks.
+        # Other scopes just don't need thread safety.
+        return
+    if not node.type.needs_refcounting:
+        # Non-refcounted types end up as thread-private anyway, so
+        # they're always safe in parallel blocks.
+        return
+    if node.cf_is_null:
+        # We know enough about it to know that no-one else has written to it
+        return
+    if not any(assignment.in_parallel for assignment in node.entry.cf_assignments):
+        return  # No assignments in a parallel region - nothing can change
+    if env._in_parallel_block:
+        node.needs_threadsafe_access = True
+        node.threadsafety_scope = entry.scope
+
+def generate_threadsafe_lock(node, code):
+    assert node.is_name or node.is_attribute
+    if not node.needs_threadsafe_access:
+        return
+    node.threadsafe_lock_error_label = code.new_error_label()
+    code.putln("{")
+    code.putln("int __pyx_critical_section_error = 1;")
```

Comment:
Essentially, the idea is:

`Py_BEGIN_CRITICAL_SECTION` and `Py_END_CRITICAL_SECTION` both add brackets, so the number of sections must match. Therefore, we can't define an extra `Py_END_CRITICAL_SECTION` on the error-handling path.

Therefore I've had to use a variable to keep track of if we're on the error handling path, and insert an extra label to always go through `Py_END_CRITICAL_SECTION`.

The same currently applies to the `PyMutex` version (which I use at function and global scope). In principle it doesn't have to though - I could separate out the paths

Reply:
That's similar to what Pyrex originally did for `try-finally`, before we moved to splitting up the `finally` clause into separate C code sections. Can't say if it's a performance regression, but I'd rather not introduce one for this.

Reply:
I think I can hack my way out of it with dummy brackets... I'll leave a comment on the issue to expose these macros publicly though and see if we can get anything anything nicer though.